diesel/src/query_builder/insert_statement.rs

```diff
@@ -148,6 +189,140 @@ impl<T, U, Op> InsertStatement<T, U, Op, NoReturningClause> {
+/// When the given slice is not empty, this will execute a single bulk insert
+/// on backends which support the `DEFAULT` keyword, and one query per record
+/// on backends which do not (SQLite).
+pub struct BatchInsertStatement<T, U, Op=Insert, Ret=NoReturningClause> {
+    operator: Op,
+    target: T,
+    records: U,
+    returning: Ret,
+}
+
+impl<T, U, Op, Ret> BatchInsertStatement<T, U, Op, Ret> {
+    fn into_insert_statement(self) -> InsertStatement<T, U, Op, Ret> {
+        InsertStatement::new(
+            self.target,
+            self.records,
+            self.operator,
+            self.returning,
+        )
+    }
+}
+
+impl<T, U, Op> BatchInsertStatement<T, U, Op> {
+    /// Specify what expression is returned after execution of the `insert`.
+    /// # Examples
+    ///
+    /// ### Inserting records:
+    ///
+    /// ```rust
+    /// # #[macro_use] extern crate diesel;
+    /// # include!("src/doctest_setup.rs");
+    /// #
+    /// # table! {
+    /// #     users {
+    /// #         id -> Integer,
+    /// #         name -> VarChar,
+    /// #     }
+    /// # }
+    /// #
+    /// # #[cfg(feature = "postgres")]
+    /// # fn main() {
+    /// #     use self::users::dsl::*;
+    /// #     let connection = establish_connection();
+    /// let new_users = vec![
+    ///     NewUser { name: "Timmy".to_string(), },
+    ///     NewUser { name: "Jimmy".to_string(), },
+    /// ];
+    ///
+    /// let inserted_names = diesel::insert(&new_users)
+    ///     .into(users)
+    ///     .returning(name)
+    ///     .get_results(&connection);
+    /// assert_eq!(Ok(vec!["Timmy".to_string(), "Jimmy".to_string()]), inserted_names);
+    /// # }
+    /// # #[cfg(not(feature = "postgres"))]
+    /// # fn main() {}
+    /// ```
+    pub fn returning<E>(self, returns: E)
+        -> BatchInsertStatement<T, U, Op, ReturningClause<E>>
+    {
+        BatchInsertStatement {
+            operator: self.operator,
+            target: self.target,
+            records: self.records,
+            returning: ReturningClause(returns),
+        }
+    }
+}
+
+impl<'a, T, U, Op, Ret, Conn, DB> ExecuteDsl<Conn, DB>
+    for BatchInsertStatement<T, &'a [U], Op, Ret> where
+        Conn: Connection<Backend=DB>,
+        DB: Backend + SupportsDefaultKeyword,
+        InsertStatement<T, &'a [U], Op, Ret>: ExecuteDsl<Conn>,
+{
+    fn execute(self, conn: &Conn) -> QueryResult<usize> {
+        if self.records.is_empty() {
+            Ok(0)
+        } else {
+            self.into_insert_statement().execute(conn)
+        }
+    }
+}
+
+#[cfg(feature="sqlite")]
+impl<'a, T, U, Op, Ret> ExecuteDsl<::sqlite::SqliteConnection>
+    for BatchInsertStatement<T, &'a [U], Op, Ret> where
+        InsertStatement<T, &'a U, Op, Ret>: ExecuteDsl<::sqlite::SqliteConnection>,
+        T: Copy,
+        Op: Copy,
+        Ret: Copy,
+{
+    fn execute(self, conn: &::sqlite::SqliteConnection) -> QueryResult<usize> {
+        let mut result = 0;
```

Comment:
I would love to use `Iterator::sum` for this, but I don't see a good way to do it without first using `Iterator::collect` to put it into a `Result<Vec<T>>`